File: adapters/ui-kotlin/src/main/kotlin/dev/flux/ui/OverlayMotionAdapters.kt
package dev.flux.ui

import java.lang.ref.WeakReference
import kotlin.collections.List as KList

/**
 * Declarative adapters for the FLUX-038 overlay containers (`Modal` / `Sheet` /
 * `Dialog`) and the FLUX-042 signal-graph animation wrapper (`Animate`),
 * unified tier (AGENTS.md §3.5).
 *
 * Each hosts its `content` / animated subtree as children and binds any
 * `onDismiss` handler. The native *presentation* (a hosted sheet / alert /
 * dialog) and the native *animation* (`withAnimation`) are gated on the ADR-0048
 * iOS dev-tier convergence decision — until then these adapters degrade to a
 * plain container carrying the children, so a Flux app can author and render the
 * primitives today without a blank screen (the dev/release parity mapping is
 * already pinned by `flux-parity`).
 *
 * Each node gets its own adapter instance via [create] (FLUX-007).
 */

/** `Modal` — centered modal over a scrim (Compose `Dialog`). */
public class ModalAdapter private constructor() : FluxAdapter<FluxNativeView> {
    override val kind: String = KIND

    override fun create(nodeId: UInt): FluxNativeView = FluxNativeViewImpl(nodeId, kind)

    override fun update(
        view: FluxNativeView,
        props: Props,
    ) {
        // Presentation is host-native (ADR-0048); we record the onDismiss handler
        // id so `bindHandler` can wire it once the native surface lands.
        val onDismiss = props.getHandler(PropsIndex.OVERLAY_ON_DISMISS)
        if (view.getProperty(PROP_ON_DISMISS) != onDismiss) view.setProperty(PROP_ON_DISMISS, onDismiss)
    }

    override fun setChildren(
        view: FluxNativeView,
        childIds: KList<UInt>,
        children: KList<FluxNativeView>,
    ) {
        reconcileChildren(view, childIds) { id -> children.firstOrNull { it.nodeId == id } }
    }

    override fun bindHandler(
        view: FluxNativeView,
        props: Props,
        executor: WeakReference<FluxExecutor>,
    ) {
        // No native event binding yet; the onDismiss callback is invoked by the
        // host presentation layer once ADR-0048 lands.
    }

    override fun destroy(view: FluxNativeView) {
        view.clearChildren()
    }

    internal companion object {
        const val KIND: String = "modal"
        const val PROP_ON_DISMISS = "onDismiss"

        fun create(): FluxAdapter<FluxNativeView> = ModalAdapter()
    }
}

/** `Sheet` — bottom-anchored sheet that slides up (Compose `ModalBottomSheet`). */
public class SheetAdapter private constructor() : FluxAdapter<FluxNativeView> {
    override val kind: String = KIND

    override fun create(nodeId: UInt): FluxNativeView = FluxNativeViewImpl(nodeId, kind)

    override fun update(
        view: FluxNativeView,
        props: Props,
    ) {
        val onDismiss = props.getHandler(PropsIndex.OVERLAY_ON_DISMISS)
        if (view.getProperty(PROP_ON_DISMISS) != onDismiss) view.setProperty(PROP_ON_DISMISS, onDismiss)
    }

    override fun setChildren(
        view: FluxNativeView,
        childIds: KList<UInt>,
        children: KList<FluxNativeView>,
    ) {
        reconcileChildren(view, childIds) { id -> children.firstOrNull { it.nodeId == id } }
    }

    override fun bindHandler(
        view: FluxNativeView,
        props: Props,
        executor: WeakReference<FluxExecutor>,
    ) {
    }

    override fun destroy(view: FluxNativeView) {
        view.clearChildren()
    }

    internal companion object {
        const val KIND: String = "sheet"
        const val PROP_ON_DISMISS = "onDismiss"

        fun create(): FluxAdapter<FluxNativeView> = SheetAdapter()
    }
}

/** `Dialog` — modal dialog with a dimmed scrim (Compose `AlertDialog`). */
public class DialogAdapter private constructor() : FluxAdapter<FluxNativeView> {
    override val kind: String = KIND

    override fun create(nodeId: UInt): FluxNativeView = FluxNativeViewImpl(nodeId, kind)

    override fun update(
        view: FluxNativeView,
        props: Props,
    ) {
        val onDismiss = props.getHandler(PropsIndex.OVERLAY_ON_DISMISS)
        if (view.getProperty(PROP_ON_DISMISS) != onDismiss) view.setProperty(PROP_ON_DISMISS, onDismiss)
    }

    override fun setChildren(
        view: FluxNativeView,
        childIds: KList<UInt>,
        children: KList<FluxNativeView>,
    ) {
        reconcileChildren(view, childIds) { id -> children.firstOrNull { it.nodeId == id } }
    }

    override fun bindHandler(
        view: FluxNativeView,
        props: Props,
        executor: WeakReference<FluxExecutor>,
    ) {
    }

    override fun destroy(view: FluxNativeView) {
        view.clearChildren()
    }

    internal companion object {
        const val KIND: String = "dialog"
        const val PROP_ON_DISMISS = "onDismiss"

        fun create(): FluxAdapter<FluxNativeView> = DialogAdapter()
    }
}

/**
 * `Animate` — signal-graph animation wrapper (FLUX-042). Hosts its child subtree
 * and records the `signal` / `curve` / `duration` data the host consumes to
 * drive the native `withAnimation` (ADR-0048). Until the native animation API is
 * wired, the children render unchanged — the node resolves (no blank screen)
 * and the curve data is carried on the view for the host layer.
 */
public class AnimateAdapter private constructor() : FluxAdapter<FluxNativeView> {
    override val kind: String = KIND

    override fun create(nodeId: UInt): FluxNativeView = FluxNativeViewImpl(nodeId, kind)

    override fun update(
        view: FluxNativeView,
        props: Props,
    ) {
        val signal = props.getHandler(PropsIndex.ANIMATE_SIGNAL)
        if (view.getProperty(PROP_SIGNAL) != signal) view.setProperty(PROP_SIGNAL, signal)
        props.getString(PropsIndex.ANIMATE_CURVE)?.let { curve ->
            if (view.getProperty(PROP_CURVE) != curve) view.setProperty(PROP_CURVE, curve)
        }
        props.getFloat(PropsIndex.ANIMATE_DURATION)?.let { duration ->
            if (view.getProperty(PROP_DURATION) != duration) view.setProperty(PROP_DURATION, duration)
        }
    }

    override fun setChildren(
        view: FluxNativeView,
        childIds: KList<UInt>,
        children: KList<FluxNativeView>,
    ) {
        reconcileChildren(view, childIds) { id -> children.firstOrNull { it.nodeId == id } }
    }

    override fun bindHandler(
        view: FluxNativeView,
        props: Props,
        executor: WeakReference<FluxExecutor>,
    ) {
    }

    override fun destroy(view: FluxNativeView) {
        view.clearChildren()
    }

    internal companion object {
        const val KIND: String = "animate"
        const val PROP_SIGNAL = "signal"
        const val PROP_CURVE = "curve"
        const val PROP_DURATION = "duration"

        fun create(): FluxAdapter<FluxNativeView> = AnimateAdapter()
    }
}

File: adapters/ui-swift/Sources/FluxUIKit/OverlayMotionAdapters.swift
//  OverlayMotionAdapters.swift
//  FluxUIKit — FLUX-038 overlay containers (Modal / Sheet / Dialog) and the
//  FLUX-042 signal-graph animation wrapper (Animate).
//
//  Declarative adapters (unified tier; AGENTS.md §3.5). Each hosts its
//  `content` / animated subtree as children and reads its data props. The
//  native *presentation* (a hosted sheet / alert / dialog) and the native
//  *animation* (`withAnimation`) are gated on the ADR-0048 iOS dev-tier
//  convergence decision — until then these adapters degrade to a plain
//  container carrying the children, so a Flux app can author and render the
//  primitives today without a blank screen (the dev/release parity mapping is
//  already pinned by `flux-parity`).

import UIKit

/// Shared base for the FLUX-038 / FLUX-042 adapters: a plain `UIView` container
/// that hosts children and records data props for the host presentation layer.
public class OverlayContainerAdapter: FluxAdapter {
    public typealias View = UIView
    weak var executor: (any FluxExecutor)?
    /// The surface name of this overlay (used by the host presentation layer
    /// once ADR-0048 lands to pick the native surface).
    public let surface: String

    public init(executor: (any FluxExecutor)? = nil, surface: String) {
        self.executor = executor
        self.surface = surface
    }

    public func create() -> UIView {
        let view = UIView()
        view.translatesAutoresizingMaskIntoConstraints = false
        return view
    }

    public func update(_ view: UIView, from old: Props, to new: Props) {
        // `onDismiss` (handler id) is recorded for parity with Android
        // `PROP_ON_DISMISS`; the host presentation layer (ADR-0048) invokes it once
        // the native surface is wired. Absent means no handler (degraded form).
        let onDismiss = new.getHandler(named: "onDismiss")
        view.fluxRecord(FluxRecordedProp.onDismiss, onDismiss as Any)
    }

    public func setChildren(_ children: [AnyObject], on view: UIView) {
        let views = children.compactMap { $0 as? UIView }
        view.subviews.forEach { $0.removeFromSuperview() }
        for v in views {
            view.addSubview(v)
            v.translatesAutoresizingMaskIntoConstraints = false
            NSLayoutConstraint.activate([
                v.leadingAnchor.constraint(equalTo: view.leadingAnchor),
                v.trailingAnchor.constraint(equalTo: view.trailingAnchor),
                v.topAnchor.constraint(equalTo: view.topAnchor),
                v.bottomAnchor.constraint(equalTo: view.bottomAnchor),
            ])
        }
    }

    public func bindHandler(_ handlerId: FluxHandlerId, to view: UIView, nodeId: FluxNodeId) {}

    public func destroy(_ view: UIView) {
        view.subviews.forEach { $0.removeFromSuperview() }
    }
}

/// `Modal` — centered modal over a scrim (SwiftUI `.fullScreenCover`).
public final class ModalAdapter: OverlayContainerAdapter {
    public init(executor: (any FluxExecutor)? = nil) { super.init(executor: executor, surface: "Modal") }
}

/// `Sheet` — bottom-anchored sheet that slides up (SwiftUI `.sheet`).
public final class SheetAdapter: OverlayContainerAdapter {
    public init(executor: (any FluxExecutor)? = nil) { super.init(executor: executor, surface: "Sheet") }
}

/// `Dialog` — modal dialog with a dimmed scrim (SwiftUI `Alert`).
public final class DialogAdapter: OverlayContainerAdapter {
    public init(executor: (any FluxExecutor)? = nil) { super.init(executor: executor, surface: "Dialog") }
}

/// `Animate` — signal-graph animation wrapper (FLUX-042). Hosts its child subtree
/// and records the `signal` / `curve` / `duration` data the host consumes to
/// drive the native `withAnimation` (ADR-0048). Until the native animation API
/// is wired, the children render unchanged — the node resolves (no blank
/// screen) and the curve data is carried on the view for the host layer.
public final class AnimateAdapter: OverlayContainerAdapter {
    public init(executor: (any FluxExecutor)? = nil) { super.init(executor: executor, surface: "Animate") }

    public override func update(_ view: UIView, from old: Props, to new: Props) {
        let signal = new.getHandler(named: "signal")
        let curve = new.getString(named: "curve")
        let duration = new.getFloat(named: "duration")
        // Recorded for parity with Android `AnimateAdapter.PROP_SIGNAL` /
        // `PROP_CURVE` / `PROP_DURATION` so the host layer (ADR-0048) drives the
        // native `withAnimation`. The children render unchanged until then.
        view.fluxRecord(FluxRecordedProp.signal, signal as Any)
        view.fluxRecord(FluxRecordedProp.curve, curve as Any)
        view.fluxRecord(FluxRecordedProp.duration, duration as Any)
    }
}
